clap_complete/src/dynamic/completer.rs

```diff
@@ -64,18 +65,96 @@ pub fn complete(
             if let Some(next_cmd) = current_cmd.find_subcommand(value) {
                 current_cmd = next_cmd;
                 pos_index = 1;
+                state = ParseState::ValueDone;
                 continue;
             }
         }
 
         if is_escaped {
             pos_index += 1;
+            state = ParseState::Pos(pos_index);
         } else if arg.is_escape() {
             is_escaped = true;
-        } else if let Some(_long) = arg.to_long() {
-        } else if let Some(_short) = arg.to_short() {
+            state = ParseState::ValueDone;
+        } else if let Some((flag, value)) = arg.to_long() {
```

Comment:
Don't we need to check the state here before any of the branches?  There is more nuance to it than that but we can fill that in over time

Reply:
Of course, we could check the state here before any of the branches, and I would prefer to do so. This way, `ParseState` can truly become a type used for state machine rather than just a type that records the current raw_arg state.

Benefits that could potentially come from this (that I can think of for now):
- Easier support for multiple values https://github.com/clap-rs/clap/issues/3921 (adding new state and state transitions).
- More precise completion (accurately completing content based on different states).

Reply:
Note: we should also have a test case for this

One way of testing this is to have an option and a positional that have different completions: `"--option not-positional positional"`

Reply:
Thanks for your reminder! There is an error for `pos_index` when parsing the command line. I just check the state in last else branch. For `arg.is_escape()`, `arg.to_long()`, `arg.to_short()`, there is a case that the `arg` may be a value rather than an option or an escape with `is_allow_hyphen_values_set` setting. It's still a bug but easy to fix! I prefer to solve this as a new issue. Should we solve this problem in the current PR or a new issue?

Reply:
We should handle `is_allow_hyphen_values_set` in a separate issue.  If one doesn't exist, can you create one?

Reply:
Okay.